File: fronted/lib/controllers/expense_controller.dart
import 'package:expense_tracker_app/models/expense_model.dart';
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import '../services/api_service.dart';
import '../models/expense_response.dart';

class ExpenseController extends GetxController {
  final ApiService _apiService = Get.find<ApiService>();

  final expenses = <ExpenseModel>[].obs;
  final isLoading = false.obs;
  final totalAmount = 0.0.obs;

  final titleController = TextEditingController();
  final amountController = TextEditingController();
  final categoryController = TextEditingController();
  final dateController = TextEditingController();

  final expenseFormKey = GlobalKey<FormState>();

  final selectedExpense = Rx<ExpenseModel?>(null);

  final selectedDate = Rx<DateTime>(DateTime.now());

  final selectedCategory = Rx<String>('');

  final categories = [
    'Food',
    'Transport',
    'Shopping',
    'Bills',
    'Entertainment',
    'Health',
    'Education',
    'Other',
  ];

  double get totalExpenses {
    return totalAmount.value > 0
        ? totalAmount.value
        : expenses.fold(0.0, (sum, expense) => sum + expense.amount);
  }

  void initializeDate() {
    dateController.text = _formatDate(selectedDate.value);
  }

  String _formatDate(DateTime date) {
    return '${date.day}/${date.month}/${date.year}';
  }

  Future<void> selectDate(BuildContext context) async {
    final DateTime? picked = await showDatePicker(
      context: context,
      initialDate: selectedDate.value,
      firstDate: DateTime(2000),
      lastDate: DateTime(2100),
    );
    if (picked != null && picked != selectedDate.value) {
      selectedDate.value = picked;
      dateController.text = _formatDate(picked);
    }
  }

  String? validateTitle(String? value) {
    if (value == null || value.isEmpty) {
      return 'Please enter expense title';
    }
    if (value.length < 2) {
      return 'Title must be at least 2 characters';
    }
    return null;
  }

  String? validateAmount(String? value) {
    if (value == null || value.isEmpty) {
      return 'Please enter amount';
    }
    final amount = double.tryParse(value);
    if (amount == null) {
      return 'Please enter a valid amount';
    }
    if (amount <= 0) {
      return 'Amount must be greater than 0';
    }
    return null;
  }

  String? validateCategory(String? value) {
    if (value == null || value.isEmpty) {
      return 'Please select a category';
    }
    return null;
  }

  String? validateDate(String? value) {
    if (value == null || value.isEmpty) {
      return 'Please select a date';
    }
    return null;
  }

  Future<void> loadExpenses() async {
    isLoading.value = true;
    try {
      final response = await _apiService.get<ExpenseListResponse>(
        '/expenses',
        requiresAuth: true,
        fromJson: (json) => ExpenseListResponse.fromJson(json),
      );

      if (response.success && response.data != null) {
        expenses.value = response.data!.expenses;
        totalAmount.value = response.data!.totalAmount;
      } else {
        Get.snackbar(
          'Error',
          response.message,
          snackPosition: SnackPosition.BOTTOM,
          backgroundColor: Get.theme.colorScheme.error,
          colorText: Get.theme.colorScheme.onError,
        );
      }
    } catch (e) {
      Get.snackbar(
        'Error',
        'Failed to load expenses: ${e.toString()}',
        snackPosition: SnackPosition.BOTTOM,
        backgroundColor: Get.theme.colorScheme.error,
        colorText: Get.theme.colorScheme.onError,
      );
    } finally {
      isLoading.value = false;
    }
  }

  Future<void> addExpense() async {
    if (expenseFormKey.currentState!.validate()) {
      isLoading.value = true;
      try {
        final category = selectedCategory.value.isNotEmpty
            ? selectedCategory.value
            : categoryController.text.trim();

        final response = await _apiService.post<ExpenseItemResponse>(
          '/expenses',
          {
            'title': titleController.text.trim(),
            'amount': double.parse(amountController.text.trim()),
            'category': category,
            'date': selectedDate.value.toIso8601String(),
          },
          requiresAuth: true,
          fromJson: (json) => ExpenseItemResponse.fromJson(json),
        );

        if (response.success && response.data != null) {
          final created = response.data!.expense;
          expenses.add(created);
          totalAmount.value += created.amount;
          clearForm();
          Get.back();

          Get.snackbar(
            'Success',
            response.data!.message ?? response.message,
            snackPosition: SnackPosition.BOTTOM,
            backgroundColor: Get.theme.colorScheme.primary,
            colorText: Get.theme.colorScheme.onPrimary,
          );
        } else {
          Get.snackbar(
            'Error',
            response.message,
            snackPosition: SnackPosition.BOTTOM,
            backgroundColor: Get.theme.colorScheme.error,
            colorText: Get.theme.colorScheme.onError,
          );
        }
      } catch (e) {
        Get.snackbar(
          'Error',
          'Failed to add expense: ${e.toString()}',
          snackPosition: SnackPosition.BOTTOM,
          backgroundColor: Get.theme.colorScheme.error,
          colorText: Get.theme.colorScheme.onError,
        );
      } finally {
        isLoading.value = false;
      }
    }
  }

  Future<void> updateExpense() async {
    if (expenseFormKey.currentState!.validate() &&
        selectedExpense.value != null) {
      isLoading.value = true;
      try {
        final category = selectedCategory.value.isNotEmpty
            ? selectedCategory.value
            : categoryController.text.trim();

        final oldAmount = selectedExpense.value!.amount;

        final response = await _apiService.put<ExpenseItemResponse>(
          '/expenses/${selectedExpense.value!.id}',
          {
            'title': titleController.text.trim(),
            'amount': double.parse(amountController.text.trim()),
            'category': category,
            'date': selectedDate.value.toIso8601String(),
          },
          requiresAuth: true,
          fromJson: (json) => ExpenseItemResponse.fromJson(json),
        );

        if (response.success && response.data != null) {
          final updated = response.data!.expense;
          final index = expenses.indexWhere(
            (e) => e.id == selectedExpense.value!.id,
          );
          if (index != -1) {
            expenses[index] = updated;
            totalAmount.value = totalAmount.value - oldAmount + updated.amount;
          }
          clearForm();
          Get.back();

          Get.snackbar(
            'Success',
            response.data!.message ?? response.message,
            snackPosition: SnackPosition.BOTTOM,
            backgroundColor: Get.theme.colorScheme.primary,
            colorText: Get.theme.colorScheme.onPrimary,
          );
        } else {
          Get.snackbar(
            'Error',
            response.message,
            snackPosition: SnackPosition.BOTTOM,
            backgroundColor: Get.theme.colorScheme.error,
            colorText: Get.theme.colorScheme.onError,
          );
        }
      } catch (e) {
        Get.snackbar(
          'Error',
          'Failed to update expense: ${e.toString()}',
          snackPosition: SnackPosition.BOTTOM,
          backgroundColor: Get.theme.colorScheme.error,
          colorText: Get.theme.colorScheme.onError,
        );
      } finally {
        isLoading.value = false;
      }
    }
  }

  Future<void> deleteExpense(String expenseId) async {
    isLoading.value = true;
    try {
      final expenseToDelete = expenses.firstWhere((e) => e.id == expenseId);
      final amountToRemove = expenseToDelete.amount;

      final response = await _apiService.delete<Map<String, dynamic>>(
        '/expenses/$expenseId',
        requiresAuth: true,
      );

      if (response.success) {
        expenses.removeWhere((expense) => expense.id == expenseId);
        totalAmount.value -= amountToRemove;

        Get.snackbar(
          'Success',
          response.message,
          snackPosition: SnackPosition.BOTTOM,
          backgroundColor: Get.theme.colorScheme.primary,
          colorText: Get.theme.colorScheme.onPrimary,
        );
      } else {
        Get.snackbar(
          'Error',
          response.message,
          snackPosition: SnackPosition.BOTTOM,
          backgroundColor: Get.theme.colorScheme.error,
          colorText: Get.theme.colorScheme.onError,
        );
      }
    } catch (e) {
      Get.snackbar(
        'Error',
        'Failed to delete expense: ${e.toString()}',
        snackPosition: SnackPosition.BOTTOM,
        backgroundColor: Get.theme.colorScheme.error,
        colorText: Get.theme.colorScheme.onError,
      );
    } finally {
      isLoading.value = false;
    }
  }

  Future<void> loadExpenseSummary() async {
    try {
      final response = await _apiService.get<ExpenseSummaryResponse>(
        '/expenses/summary',
        requiresAuth: true,
        fromJson: (json) => ExpenseSummaryResponse.fromJson(json),
      );

      if (response.success && response.data != null) {
        totalAmount.value = response.data!.totalAmount;
      }
    } catch (e) {
      print('Failed to load summary: $e');
    }
  }

  void initializeFormForEdit(ExpenseModel expense) {
    selectedExpense.value = expense;
    titleController.text = expense.title;
    amountController.text = expense.amount.toString();
    categoryController.text = expense.category;
    selectedCategory.value = expense.category;
    selectedDate.value = expense.date;
    dateController.text = _formatDate(expense.date);
  }

  void clearForm() {
    titleController.clear();
    amountController.clear();
    categoryController.clear();
    selectedCategory.value = '';
    selectedDate.value = DateTime.now();
    dateController.text = _formatDate(DateTime.now());
    selectedExpense.value = null;
  }

  @override
  void onInit() {
    super.onInit();
    initializeDate();
    loadExpenses();
  }

  @override
  void onClose() {
    titleController.dispose();
    amountController.dispose();
    categoryController.dispose();
    dateController.dispose();
    super.onClose();
  }
}

File: backend/controllers/expenseController.js
const Expense = require('../models/Expense');

exports.getExpenses = async (req, res) => {
  try {
    const expenses = await Expense.find({ user: req.user._id })
      .sort({ date: -1 })
      .select('-__v');

    const totalAmount = expenses.reduce((sum, expense) => sum + expense.amount, 0);

    res.status(200).json({
      success: true,
      count: expenses.length,
      totalAmount,
      expenses
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error fetching expenses'
    });
  }
};

exports.getExpense = async (req, res) => {
  try {
    const expense = await Expense.findOne({
      _id: req.params.id,
      user: req.user._id
    }).select('-__v');

    if (!expense) {
      return res.status(404).json({
        success: false,
        message: 'Expense not found'
      });
    }

    res.status(200).json({
      success: true,
      expense
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error fetching expense'
    });
  }
};

exports.createExpense = async (req, res) => {
  try {
    const { title, amount, category, date } = req.body;

    if (!title || !amount || !category) {
      return res.status(400).json({
        success: false,
        message: 'Please provide title, amount, and category'
      });
    }

    const expense = await Expense.create({
      title,
      amount: parseFloat(amount),
      category,
      date: date ? new Date(date) : new Date(),
      user: req.user._id
    });

    res.status(201).json({
      success: true,
      message: 'Expense created successfully',
      expense
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error creating expense'
    });
  }
};

exports.updateExpense = async (req, res) => {
  try {
    const { title, amount, category, date } = req.body;

    let expense = await Expense.findOne({
      _id: req.params.id,
      user: req.user._id
    });

    if (!expense) {
      return res.status(404).json({
        success: false,
        message: 'Expense not found'
      });
    }

    if (title) expense.title = title;
    if (amount !== undefined) expense.amount = parseFloat(amount);
    if (category) expense.category = category;
    if (date) expense.date = new Date(date);

    await expense.save();

    res.status(200).json({
      success: true,
      message: 'Expense updated successfully',
      expense
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error updating expense'
    });
  }
};

exports.deleteExpense = async (req, res) => {
  try {
    const expense = await Expense.findOne({
      _id: req.params.id,
      user: req.user._id
    });

    if (!expense) {
      return res.status(404).json({
        success: false,
        message: 'Expense not found'
      });
    }

    await Expense.findByIdAndDelete(req.params.id);

    res.status(200).json({
      success: true,
      message: 'Expense deleted successfully'
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error deleting expense'
    });
  }
};

exports.getSummary = async (req, res) => {
  try {
    const expenses = await Expense.find({ user: req.user._id });

    const totalAmount = expenses.reduce((sum, expense) => sum + expense.amount, 0);

    const categoryBreakdown = expenses.reduce((acc, expense) => {
      const category = expense.category;
      if (!acc[category]) {
        acc[category] = 0;
      }
      acc[category] += expense.amount;
      return acc;
    }, {});

    res.status(200).json({
      success: true,
      summary: {
        totalExpenses: expenses.length,
        totalAmount,
        categoryBreakdown
      }
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      message: error.message || 'Error fetching summary'
    });
  }
};

